src/carma-platform/parse_ros_bags.py

```diff
@@ -0,0 +1,94 @@
+import argparse
+import csv
+from pathlib import Path
+
+import more_itertools
+import rosbag
+
+
+def find_msg_closest_to_timestamp(bag_msgs, timestamp):
+    for current, next_ in more_itertools.pairwise(bag_msgs):
+        # Each element in `bag_msgs` is a triple:
+        # <topic_name, msg, system_time_at_receive>
+        _, _, current_log_timestamp = current
+        _, _, next_log_timestamp = next_
+
+        # We only need to iterate until we find two messages who's logged
+        # timestamps straddle the timestamp of interest. All messages
+        # afterwards will have increasing differences.
+        if next_log_timestamp > timestamp:
+            if abs(timestamp - current_log_timestamp) < abs(next_log_timestamp - timestamp):
+                return current
+
+            return next_
+
+
+# calculate the offset between ros clock and the sim clock
+def get_time_offset(ros_bag_file):
+    with rosbag.Bag(ros_bag_file, "r") as bag:
+        cdasim_clock_msgs = bag.read_messages(topics=["/sim_clock"])  # CDASim clock
+        carla_clock_msgs = bag.read_messages(topics=["/clock"])  # CARLA clock
+
+        _, cdasim_clock_msg, cdasim_clock_log_time = next(cdasim_clock_msgs)
+
+        carla_clock_at_cdasim_start = find_msg_closest_to_timestamp(
+            carla_clock_msgs, cdasim_clock_log_time
+        )
+
+        return carla_clock_at_cdasim_start[1].clock - cdasim_clock_msg.clock
```

Comment:
I thought we want to reference same index.

Reply:
`carla_clock_at_cdasim_start` is a triple of `<topic, msg, log_time>`. This is grabbing the `msg` component.

Reply:
Makes sense